### PR TITLE
[Etcm 521] Fast sync integration tests

### PR DESCRIPTION
**Pull request by robinraju**
# Description

This PR adds integration tests for the newly introduced fast sync improvements (Concurrent download of headers)

# Important Changes Introduced

Minor fixes were applied to the original implementation of `FastSync`, which prevented the branch resolution from being completed properly.

# Testing

Run the tests and look through the logs for branch resolution when a malicious master peer is found.

### CODE REVIEW

File: src/main/scala/io/iohk/ethereum/blockchain/sync/fast/FastSync.scala [hunk 1]
```diff
@@ -346,13 +346,13 @@ class FastSync(
...
       }
 
-      syncState = syncState.enqueueHeaderRange(request)
       error match {
         // These are the reasons that make the master peer suspicious
         case InvalidPenultimateHeader(_, header) => handleMasterPeerFailure(header)
         case InvalidBatchHash(_, header) => handleMasterPeerFailure(header)
         // Otherwise probably it's just this peer's fault
         case _ =>
+          syncState = syncState.enqueueHeaderRange(request)
```
Comment (robinraju) [inline src/main/scala/io/iohk/ethereum/blockchain/sync/fast/FastSync.scala]: The current failed request (for headers) is not required to be enqueued if the master peer is invalid. We are discarding those and a fresh request is created based on the result obtained from BranchResolver. Hence this change.

File: src/it/scala/io/iohk/ethereum/sync/FastSyncItSpec.scala [hunk 2]
```diff
@@ -180,6 +182,65 @@ class FastSyncItSpec extends FlatSpecBase with Matchers with BeforeAndAfterAll {
...
+      // Peer3 is blacklisted
+      val blacklistedPeer = PeerId(s"${peer3.node.addr.getHostAddress}:${peer3.node.tcpPort}")
+      val blacklistReason = peer1.blacklist.cache.getIfPresent(blacklistedPeer)
+
+      assert(peer1.blacklist.isBlacklisted(blacklistedPeer))
+      assert(blacklistReason.get == BlacklistReasonType.BlockHeaderValidationFailedType)
+    }
+  }
+
+  it should "start branch resolving on invalid batch last header hash" in customTestCaseResourceM(
```
Comment (AnastasiiaL) [inline src/it/scala/io/iohk/ethereum/sync/FastSyncItSpec.scala]: should the test scenario be renamed? Or how do we check that  branch resolving happened?

Comment (robinraju) [reply src/it/scala/io/iohk/ethereum/sync/FastSyncItSpec.scala]: The occurrence of branch resolving can only be identified from logs. Maybe renaming the test is better.

File: src/it/scala/io/iohk/ethereum/sync/util/CommonFakePeer.scala [hunk 3]
```diff
@@ -298,21 +300,88 @@ abstract class CommonFakePeer(peerName: String, fakePeerCustomConfig: FakePeerCu
...
+          )
+        )
+      val newWeight = ChainWeight.totalDifficultyOnly(1)
+
+      broadcastBlock(childBlock, newWeight)
+      bl.save(childBlock, Seq(), newWeight, saveAsBestBlock = true)
+    }
+  }
+
+  private def generateValidBlocks(
```
Comment (AnastasiiaL) [inline src/it/scala/io/iohk/ethereum/sync/util/CommonFakePeer.scala]: guess this should singular - generateValidBlock

File: src/it/scala/io/iohk/ethereum/sync/util/CommonFakePeer.scala [hunk 4]
```diff
@@ -298,21 +300,88 @@ abstract class CommonFakePeer(peerName: String, fakePeerCustomConfig: FakePeerCu
     (newBlock, newWeight, parentWorld)
   }
 
+  private def generateInvalidBlocks(
```
Comment (AnastasiiaL) [inline src/it/scala/io/iohk/ethereum/sync/util/CommonFakePeer.scala]: and here also singular - generateInvalidBlock

File: src/it/scala/io/iohk/ethereum/sync/FastSyncItSpec.scala [hunk 5]
```diff
@@ -180,6 +182,65 @@ class FastSyncItSpec extends FlatSpecBase with Matchers with BeforeAndAfterAll {
     }
   }
 
+  it should "switch to regular sync once `safeDownloadTarget` is reached" in customTestCaseResourceM(
```
Comment (jvdp) [inline src/it/scala/io/iohk/ethereum/sync/FastSyncItSpec.scala:185]: Where do you assert that regular sync was initiated?

File: src/it/scala/io/iohk/ethereum/sync/util/CommonFakePeer.scala [hunk 6]
```diff
@@ -298,21 +300,88 @@ abstract class CommonFakePeer(peerName: String, fakePeerCustomConfig: FakePeerCu
     (newBlock, newWeight, parentWorld)
   }
 
+  private def generateInvalidBlock(
```
Comment (jvdp) [inline src/it/scala/io/iohk/ethereum/sync/util/CommonFakePeer.scala:303]: Perhaps put a comment where / how this block is made invalid?

Also, below is a method that mentions "MaliciousBlock", is this the same thing? If so, would be better to keep a consistent name.

**Review by jvdp — APPROVED** [review]
LGTM 👍 

Did raise some questions, and I think more comments/descriptions would be nice to have for `FakePeer` DSL.

**Review by milenkara — APPROVED** [review]

File: src/it/scala/io/iohk/ethereum/sync/FastSyncItSpec.scala (hunk 5 above)
Comment (robinraju) [reply src/it/scala/io/iohk/ethereum/sync/FastSyncItSpec.scala:185]: the `CommonFakePeer#waitForFastSyncFinish()` helper method checks for the completion of Fastsync. More specific assertion can be made possible by intercepting the logs only. Maybe something to improve in another ticket to add support for intercepting logs in tests.

**Pull request merged**